src/Model/GServer.php

```diff
@@ -0,0 +1,1148 @@
+			$server = self::parseNodeinfo2($nodeinfo2_url);
+		}
+
+		// When the nodeinfo url isn't on the same host, then there is obviously something wrong
+		if (empty($server) && !empty($nodeinfo1_url) && (parse_url($url, PHP_URL_HOST) == parse_url($nodeinfo1_url, PHP_URL_HOST))) {
+			$server = self::parseNodeinfo1($nodeinfo1_url);
+		}
+
+		return $server;
+	}
+
+	/**
+	 * Parses Nodeinfo 1
+	 *
+	 * @param string $nodeinfo_url address of the nodeinfo path
+	 * @return array Server data
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	private static function parseNodeinfo1(string $nodeinfo_url)
+	{
+		$curlResult = Network::curl($nodeinfo_url);
+
+		if (!$curlResult->isSuccess()) {
+			return false;
+		}
+
+		$nodeinfo = json_decode($curlResult->getBody(), true);
+
+		if (!is_array($nodeinfo)) {
+			return false;
+		}
+
+		$server = [];
+
+		$server['register_policy'] = Register::CLOSED;
+
+		if (!empty($nodeinfo['openRegistrations'])) {
+			$server['register_policy'] = Register::OPEN;
+		}
+
+		if (is_array($nodeinfo['software'])) {
+			if (!empty($nodeinfo['software']['name'])) {
+				$server['platform'] = $nodeinfo['software']['name'];
+			}
+
+			if (!empty($nodeinfo['software']['version'])) {
+				$server['version'] = $nodeinfo['software']['version'];
+				// Version numbers on Nodeinfo are presented with additional info, e.g.:
+				// 0.6.3.0-p1702cc1c, 0.6.99.0-p1b9ab160 or 3.4.3-2-1191.
+				$server['version'] = preg_replace('=(.+)-(.{4,})=ism', '$1', $server['version']);
+			}
+		}
+
+		if (!empty($nodeinfo['metadata']['nodeName'])) {
+			$server['site_name'] = $nodeinfo['metadata']['nodeName'];
+		}
+
+		if (!empty($nodeinfo['usage']['users']['total'])) {
+			$server['registered-users'] = $nodeinfo['usage']['users']['total'];
+		}
+
+		if (!empty($nodeinfo['protocols']['inbound']) && is_array($nodeinfo['protocols']['inbound'])) {
+			$protocols = [];
+			foreach ($nodeinfo['protocols']['inbound'] as $protocol) {
+				$protocols[$protocol] = true;
+			}
+
+			if (!empty($protocols['friendica'])) {
+				$server['network'] = Protocol::DFRN;
+			} elseif (!empty($protocols['activitypub'])) {
+				$server['network'] = Protocol::ACTIVITYPUB;
+			} elseif (!empty($protocols['diaspora'])) {
+				$server['network'] = Protocol::DIASPORA;
+			} elseif (!empty($protocols['ostatus'])) {
+				$server['network'] = Protocol::OSTATUS;
+			} elseif (!empty($protocols['gnusocial'])) {
+				$server['network'] = Protocol::OSTATUS;
+			} elseif (!empty($protocols['zot'])) {
+				$server['network'] = Protocol::ZOT;
+			}
+		}
+
+		if (!$server) {
+			return false;
+		}
+
+		return $server;
+	}
+
+	/**
+	 * Parses Nodeinfo 2
+	 *
+	 * @param string $nodeinfo_url address of the nodeinfo path
+	 * @return array Server data
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	private static function parseNodeinfo2(string $nodeinfo_url)
```

Comment:
There is a bunch of duplicated code between this method and `parseNodeinfo1`. Ideally, we should have a `Protocol\NodeInfoV2` class extending `Protocol\NodeInfoV1` that would include the common extraction methods to both versions.

Reply:
I hadn't found a good way that wouldn't reduce the readability.